File: Lab06Cs/Lab06Monitoring/TemperatureSensor.cs
using System;

namespace Lab06Monitoring
{
    /// <summary>
    /// Датчик температуры.
    ///
    /// ЛР-6:
    ///  * переопределяем виртуальные методы;
    ///  * в Calibrate() вызываем базовую реализацию;
    ///  * реализуем Clone() (поверхностное клонирование);
    ///  * объявляем метод TypeName() с той же сигнатурой, что и в базовом классе,
    ///    но без virtual.
    /// </summary>
    public class TemperatureSensor : Sensor
    {
        public float MinThreshold { get; }
        public float MaxThreshold { get; }

        private static readonly Random _rnd = new();

        public TemperatureSensor(string id, string name, float min, float max)
            : base(id, name, "°C")
        {
            if (min >= max)
                throw new ArgumentException("min must be < max", nameof(min));

            MinThreshold = min;
            MaxThreshold = max;
        }

        public override float ReadValue()
        {
            LastValue = (float)(_rnd.NextDouble() * (MaxThreshold - MinThreshold + 20) + (MinThreshold - 10));
            Console.WriteLine($"TemperatureSensor '{Name}' read: {LastValue:F2} {Unit}");
            return LastValue;
        }

        public override bool CheckStatus()
        {
            bool ok = LastValue >= MinThreshold && LastValue <= MaxThreshold;
            Console.WriteLine($"TemperatureSensor '{Name}' status: {(ok ? "OK" : "OUT OF RANGE")}");
            return ok;
        }

        public override void Calibrate()
        {
            Console.WriteLine($"TemperatureSensor '{Name}': custom calibration...");
            base.Calibrate(); // вызов базового метода
        }

        public override string Describe()
        {
            return $"TemperatureSensor[{Id}] '{Name}', range = {MinThreshold}..{MaxThreshold} {Unit}";
        }

        public new string TypeName()
        {
            // new скрывает базовую реализацию (не полиморфно)
            return "TemperatureSensor";
        }

        public override object Clone()
        {
            // Поверхностное клонирование: копируем простые поля.
            return new TemperatureSensor(Id, Name, MinThreshold, MaxThreshold)
            {
                LastValue = this.LastValue,
                Status = this.Status
            };
        }
    }
}

File: Lab08Cpp/TemperatureSensor.cpp
#include "TemperatureSensor.hpp"
#include <stdexcept>
#include <iomanip>

TemperatureSensor::TemperatureSensor(const std::string& id,
                                     const std::string& name,
                                     float min,
                                     float max)
    : Sensor(id, name, "°C")
    , m_min(min)
    , m_max(max)
{
    if (min >= max)
    {
        throw std::invalid_argument("TemperatureSensor: min threshold must be < max threshold");
    }
}

float TemperatureSensor::ReadValue()
{
    // Простая генерация псевдослучайной температуры.
    static std::mt19937 gen{std::random_device{}()};
    std::uniform_real_distribution<float> dist(m_min - 10.0f, m_max + 10.0f);

    m_lastValue = dist(gen);
    std::cout << "TemperatureSensor '" << m_name
              << "' read value: " << std::fixed << std::setprecision(2)
              << m_lastValue << " " << m_unit << std::endl;
    return m_lastValue;
}

bool TemperatureSensor::CheckStatus() const
{
    bool ok = (m_lastValue >= m_min && m_lastValue <= m_max);
    std::cout << "TemperatureSensor '" << m_name << "' status: "
              << (ok ? "OK" : "OUT OF RANGE") << std::endl;
    return ok;
}

void TemperatureSensor::Calibrate()
{
    // ЛР-6: пример переопределения метода базового класса с вызовом base-реализации.
    std::cout << "TemperatureSensor '" << m_name << "': custom calibration..." << std::endl;
    Sensor::Calibrate(); // Вызов базового варианта.
}

std::string TemperatureSensor::Describe() const
{
    return "TemperatureSensor[" + m_id + "] '" + m_name + "', range = " +
           std::to_string(m_min) + ".." + std::to_string(m_max) + " " + m_unit;
}

Sensor* TemperatureSensor::Clone() const
{
    // Поверхностное клонирование по данным: здесь оно эквивалентно копированию,
    // так как нет вложенных указателей.
    return new TemperatureSensor(*this);
}

TemperatureSensor& TemperatureSensor::operator=(const Sensor& base)
{
    if (this != &base)
    {
        // ЛР-6: пример присваивания объекту производного класса
        // состояния объекта базового класса.
        m_id     = base.GetId();
        m_name   = base.GetName();
        m_unit   = base.GetUnit();
        m_lastValue = base.GetLastValue();
        m_status = base.GetStatus();

        // Диапазон температур оставим прежним (или можно задать по умолчанию).
    }
    return *this;
}

std::string TemperatureSensor::TypeName() const
{
    return "TemperatureSensor";
}
